### PR TITLE
Add flow elements

### PR DESCRIPTION
**Pull request by HimashiRathnayake**
## Purpose
This PR adds the flow elements(nodes and edges) required in the visual flow of the Authentication Flow Composer.

## Goals
Adds the flow elements required in the visual flow.

### CODE REVIEW

File: src/components/visual-flow/edges/CustomEdge.tsx [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { FunctionComponent, ReactElement } from "react";
+import { ArrowHeadType, EdgeText, Position, getEdgeCenter, getMarkerEnd, getSmoothStepPath } from "react-flow-renderer";
```
Comment (brionmario) [inline src/components/visual-flow/edges/CustomEdge.tsx]:
```suggestion
import {
    ArrowHeadType,
    EdgeText,
    Position,
    getEdgeCenter,
    getMarkerEnd,
    getSmoothStepPath
} from "react-flow-renderer";
```

File: src/components/visual-flow/edges/CustomEdge.tsx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+ * Custom Edge component.
+ *
+ * @param {CustomEdgeProps} props - Props injected to the component.
+ *
+ * @return {React.ReactElement}
+ */
+export const CustomEdge : FunctionComponent<CustomEdgeProps> = (props: CustomEdgeProps) : ReactElement => {
+
+    const {
+        id, sourceX, sourceY, targetX, targetY, sourcePosition, targetPosition, data, arrowHeadType, markerEndId
```
Comment (brionmario) [inline src/components/visual-flow/edges/CustomEdge.tsx]:
```suggestion
        id,
        sourceX,
        sourceY,
        targetX,
        targetY,
        sourcePosition,
        targetPosition,
        data,
        arrowHeadType,
        markerEndId
```

File: src/components/visual-flow/edges/CustomEdge.tsx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    } = props;
+
+    let offset: number = data.offset,
```
Comment (brionmario) [inline src/components/visual-flow/edges/CustomEdge.tsx]: Let's declare these variables seperately.
```suggestion
    let offset: number = data.offset,
```

File: src/components/visual-flow/nodes/authentication-option-node.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+import Jwt from "../../../assets/jwt.png";
+import { Facebook, Fido, Google, Instagram, Totp, Twitter } from "../../core";
+
+/**
+ * Authentication Option Node component.
+ *
+ * @return {React.ReactElement}
+ */
+// eslint-disable-next-line react/prop-types,@typescript-eslint/ban-ts-comment
+// @ts-ignore
```
Comment (brionmario) [inline src/components/visual-flow/nodes/authentication-option-node.tsx]: What is the reason for `@ts-ignore` here. It is usually an anti pattern.

File: src/components/visual-flow/nodes/step-node.tsx [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+/**
+ * Step Node component.
+ *
+ * @return {React.ReactElement}
+ */
+// eslint-disable-next-line react/prop-types,@typescript-eslint/ban-ts-comment
+// @ts-ignore
+export const StepNode: FunctionComponent = ({ data }) : ReactElement => {
+
+
```
Comment (brionmario) [inline src/components/visual-flow/nodes/step-node.tsx:46]:
```suggestion
```

File: src/components/visual-flow/edges/custom-edge.tsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        arrowHeadType,
+        markerEndId
+    } = props;
+
+    let offset: number = data.offset;
+    let targetOffsetX: number = data.targetOffsetX;
+    let targetOffsetY: number = data.targetOffsetY;
+    let middleOffset: number = data.middleOffset;
+    let borderRadius: number = 0;
+    if (offset===undefined) offset = 0;
```
Comment (brionmario) [inline src/components/visual-flow/edges/custom-edge.tsx:107]:
```suggestion
   
    if (offset===undefined) offset = 0;
```

File: src/components/visual-flow/nodes/failure-node.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            </div>
+            <Handle
+                className="hidden-handle"
+                id = "targetBottom"
+                type="target"
+                position={ Position.Bottom }
+            />
+        </div>
+    );
+
```
Comment (brionmario) [inline src/components/visual-flow/nodes/failure-node.tsx]:
```suggestion
```

File: src/components/visual-flow/nodes/done-node.tsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+                position={ Position.Top }
+                id="targetTop"
+            />
+            <div>
+                <AiOutlineCheckCircle/>
+            </div>
+            <div className="done-node-bottom-text">Done</div>
+        </div>
+    );
+
```
Comment (brionmario) [inline src/components/visual-flow/nodes/done-node.tsx]:
```suggestion
```

File: src/components/visual-flow/nodes/condition-node.tsx [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            />
+            <Handle
+                className="hidden-handle"
+                type="source"
+                position={ Position.Top }
+                id="sourceTop"
+            />
+        </div>
+    );
+
```
Comment (brionmario) [inline src/components/visual-flow/nodes/condition-node.tsx]:
```suggestion
```

File: src/components/visual-flow/nodes/authentication-option-node.tsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            }
+            <Handle
+                className="hidden-handle"
+                type="source"
+                position={ Position.Right }
+                id="sourceRight"
+            />
+        </div>
+    );
+
```
Comment (brionmario) [inline src/components/visual-flow/nodes/authentication-option-node.tsx]:
```suggestion
```

File: src/components/visual-flow/flow-elements.tsx [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+            fontSize: 14,
+            letterSpacing:"1px"
+        },
+        source: source,
+        sourceHandle: handler,
+        target: target,
+        targetHandle: targetHandler,
+        type: "step"
+    });
+
```
Comment (brionmario) [inline src/components/visual-flow/flow-elements.tsx]:
```suggestion
```

File: src/components/visual-flow/flow-elements.tsx [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+ * @return {FlowElement}
+ */
+export const FailureElement = (id:string, x:number, y:number) : FlowElement => {
+
+    return ({
+        id: id,
+        position: { x: x, y: y },
+        type: "failure"
+    });
+
```
Comment (brionmario) [inline src/components/visual-flow/flow-elements.tsx]:
```suggestion
```

File: src/components/visual-flow/flow-elements.tsx [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+            text : label
+        },
+        id: id,
+        source: source,
+        sourceHandle: handler,
+        target: target,
+        targetHandle: targetHandler,
+        type: "custom"
+    });
+
```
Comment (brionmario) [inline src/components/visual-flow/flow-elements.tsx]:
```suggestion
```

File: src/components/visual-flow/flow-elements.tsx [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+ * @return {FlowElement}
+ */
+export const StartElement = (id:string, x:number, y:number) : FlowElement => {
+
+    return ({
+        id: id,
+        position: { x: x, y: y },
+        type: "start"
+    });
+
```
Comment (brionmario) [inline src/components/visual-flow/flow-elements.tsx]:
```suggestion
```

File: src/components/visual-flow/flow-elements.tsx [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+    return ({
+        data: {
+            type: type
+        },
+        id: id,
+        position: { x: x, y: y },
+        type: "authenticationOption"
+    });
+
```
Comment (brionmario) [inline src/components/visual-flow/flow-elements.tsx]:
```suggestion
```

File: src/components/visual-flow/flow-elements.tsx [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+ * @return {FlowElement}
+ */
+export const PlusElement = (id:string, x:number, y:number) : FlowElement => {
+
+    return ({
+        id: id,
+        position: { x: x, y: y },
+        type: "plus"
+    });
+
```
Comment (brionmario) [inline src/components/visual-flow/flow-elements.tsx]:
```suggestion
```

File: src/components/visual-flow/flow-elements.tsx [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+ * @return {FlowElement}
+ */
+export const DoneElement = (id:string, x:number, y:number) : FlowElement => {
+
+    return ({
+        id: id,
+        position: { x: x, y: y },
+        type: "done"
+    });
+
```
Comment (brionmario) [inline src/components/visual-flow/flow-elements.tsx]:
```suggestion
```

File: src/components/visual-flow/flow-elements.tsx [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+    return ({
+        data: {
+            args: args,
+            condition: condition
+        },
+        id: id,
+        position: { x: x, y: y },
+        type: "condition"
+    });
+
```
Comment (brionmario) [inline src/components/visual-flow/flow-elements.tsx]:
```suggestion
```

File: src/components/visual-flow/flow-elements.tsx [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+        data: {
+            onClick: onClick,
+            showAuthenticatorsList: showAuthenticatorsList,
+            text: step
+        },
+        id: step,
+        position: { x: x, y: y },
+        type: "authenticationStep"
+    });
+
```
Comment (brionmario) [inline src/components/visual-flow/flow-elements.tsx]:
```suggestion
```

File: src/components/visual-flow/edges/custom-edge.tsx [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+                    {
+                        fill:"#fff",
+                        fontSize: 14,
+                        letterSpacing:"1px"
+                    }
+                }
+            />
+        </g>
+    );
+
```
Comment (brionmario) [inline src/components/visual-flow/edges/custom-edge.tsx]:
```suggestion
```

**Review by brionmario — APPROVED** [review]

File: src/components/visual-flow/nodes/authentication-option-node.tsx (hunk 4 above)
Comment (HimashiRathnayake) [reply src/components/visual-flow/nodes/authentication-option-node.tsx]: Fixed with [34d0251](https://github.com/wso2-incubator/identity-authentication-flow-composer/pull/8/commits/34d0251e3a545bb19ec6307d398fe375a97d2b52)

File: src/components/visual-flow/nodes/step-node.tsx (hunk 5 above)
Comment (HimashiRathnayake) [reply src/components/visual-flow/nodes/step-node.tsx:46]: Fixed with [3548626](https://github.com/wso2-incubator/identity-authentication-flow-composer/pull/8/commits/3548626908643b65bd20372002c64d56a1c6ec39)

**Pull request merged**